controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go

```diff
@@ -310,22 +316,24 @@ func (r *KubeadmControlPlaneReconciler) updateStatus(ctx context.Context, kcp *c
 	return nil
 }
 
-func (r *KubeadmControlPlaneReconciler) scaleUpControlPlane(ctx context.Context, cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane, numMachines int) error {
-	var errs []error
+func (r *KubeadmControlPlaneReconciler) scaleUpControlPlane(ctx context.Context, cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane, ownedMachines *clusterv1.MachineList) error {
+	cli := etcd.Client{}
+	if err := cli.CheckAllStarted(); err != nil {
```

Comment:
Would it be the responsibility of `CheckAllStarted()` to ensure that each expected member (the instance backed by each ownedMachine) is considered a member of the etcd Cluster? Or should we expect that to be handled by `CheckClusterHealth` or possibly a separate test?

Reply:
Great point!

For example, if some ownedMachine is not in the list of members, that is unexpected and indicates the machine failed to bootstrap correctly, or that it bootstrapped correctly, but something removed the member afterward.

Conversely, if there is an etcd member for which there is no ownedMachine, that member is not  being operated by the controller, which is unexpected and indicates an error somewhere.

Let me give this some thought.

Reply:
In some of the cases you mentioned we may want to reconcile the issue for the purposes of re-entrancy, but I think we definitely need a way to detect the situation so that we can handle it appropriately.

Reply:
You're right.

Just after being created, the ownedMachine will _not_ be in the list of members, so I was incorrect to say that it's unexpected. The expected progression for an ownedMachine is:
1. Not in the list of members (before kubeadm makes "add member" etcd API call)
1. In the list of members, unstarted (after kubeadm makes "add member" etcd API call)
1. In the list of members, started (after etcd process in the static pod starts and reaches its peers)

--

Would you agree with this:

> The controller should create a new replica only if,  for every ownedMachine, there is a healthy etcd member (which implies that the member is listed and started).

Reply:
How does the controller identify the etcd member for an ownedMachine?

It must rely on a piece of information that is both unique for the pair, and that can be read both from an ownedMachine and an etcd member entry.

Every etcd member has a unique name (this is defined by the user, and is not the member ID).

For a machine created with CABPK, the etcd member _Name_ is equal to the kubeadm configuration `NodeRegistration.Name`. (see the kubeadm code [here](https://github.com/kubernetes/kubernetes/blob/v1.17.2/cmd/kubeadm/app/cmd/phases/join/controlplanejoin.go#L143) and [here](https://github.com/kubernetes/kubernetes/blob/v1.17.2/cmd/kubeadm/app/cmd/phases/init/etcd.go#L96))

However, the name is defined only after the etcd member starts.

So, if the controller can iterate over the list of ownedMachines. For every ownedMachine, it will look for an etcd member with a matching Name. If the controller cannot find a member with a matching Name, then the member has either not been added to the list, or it has not started. In either case, the controller should return an error.  If the controller finds a member with a matching Name, then it should verify that each of that member's endpoints (i.e. client URLs) is healthy (i.e. it completes an etcd Get request).

We may also want the controller to verify that the etcd member reports no Alarms.

>  If the keyspace's backend database for any member exceeds the space quota, etcd raises a cluster-wide alarm that puts the cluster into a maintenance mode which only accepts key reads and deletes.

(I take this to mean that using a Get to check member health will miss the fact that the cluster cannot accept a Put)